In [4]:
import os
os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID" 
device = 2
os.environ["CUDA_VISIBLE_DEVICES"]= str(device)
import torch 
import torch.nn as nn
device = torch.device("cuda:" + str(device) if torch.cuda.is_available() else "cpu")

import pandas as pd
import time
from multiprocessing import Pool
import os
import re
import numpy as np
import pycld2 as cld2  # language recognition 
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from collections import Counter
# Let Jupyter Notebook show more content under cell
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

In [5]:
import pandas as pd

## 1. Read in the raw unlabeled data

In [6]:
df = pd.read_json('/shared/3/projects/benlitterer/podcastData/processed/floydMonth/floydMonthData.jsonl', orient="records", lines=True)

In [7]:
df.head(1)

potentialOutPath  \
0  /www.buzzsprout.com/3y/httpswww.buzzsprout.com7830203983942todayisyesterday.mp3MERGED   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               

In [5]:
soted_dates = sorted(df['pubDate'])

In [17]:
df.columns

Index(['potentialOutPath', 'transcript', 'wCount', 'rssUrl', 'epTitle',
       'epDescription', 'duration', 'pubDate', 'copyright', 'itunes:type',
       'itunes:complete', 'guid', 'itunes:explicit', 'enclosure',
       'itunes:image', 'transDict', 'id', 'title', 'lastUpdate', 'link',
       'lastHttpStatus', 'dead', 'contentType', 'itunesId', 'originalUrl',
       'itunesAuthor', 'itunesOwnerName', 'explicit', 'imageUrl', 'itunesType',
       'generator', 'newestItemPubdate', 'language', 'oldestItemPubdate',
       'episodeCount', 'popularityScore', 'priority', 'createdOn',
       'updateFrequency', 'chash', 'host', 'newestEnclosureUrl', 'podcastGuid',
       'podDescription', 'category1', 'category2', 'category3', 'category4',
       'category5', 'category6', 'category7', 'category8', 'category9',
       'category10', 'newestEnclosureDuration', 'oldestItemDatetime',
       'cleanDates', 'cleanDatesLoc', 'parsedDuration', 'hours', 'transEnts',
       'transStarts', 'transEnds', 'trans

In [ ]:
df.shape

## 2. Labeling strategy: 
### - A podcast is "politics" if it only has one single category assigned, and that being "politics". If none of its 10 category has "politics", it's non-politics
### - As we rely on weakly labeled data, this can keep our labeled data as clean as possible


In [25]:
def label_pol(row):
    if str(row['category1']).lower() == 'politics' and all(pd.isnull(row[col]) for col in ['category2', 'category3', 'category4','category5','category6','category7','category8','category9', 'category10']):
        return 1
    elif not any('politics' in str(category).lower() for category in row[['category1', 'category2', 'category3', 'category4', 'category5', 'category6', 'category7', 'category8', 'category9', 'category10']]):
        return 0
    else:
        return 2

In [ ]:
pd.options.mode.chained_assignment = None
df['is_politics'] = df.apply(label_pol, axis=1)

In [ ]:
df[df["is_politics"]==0].shape

In [ ]:
df[df["is_politics"]==1].shape

## The first method yielded too limited positive samples. Thus, we label a podcast as politics as long as the top 1 category is "politics"

In [8]:
def label_pol_new(row):
    if str(row['category1']).lower() == 'politics':
        return 1
    elif not any('politics' in str(category).lower() for category in row[['category1', 'category2', 'category3', 'category4', 'category5', 'category6', 'category7', 'category8', 'category9', 'category10']]):
        return 0
    else:
        return 2

In [9]:
pd.options.mode.chained_assignment = None
df['is_politics_new'] = df.apply(label_pol_new, axis=1)

In [13]:
df[df["is_politics_new"]==1].shape

(22, 74)

## Also too limited data. Try the third strategy. We label a podcast as politics if its top 3 category includes "politics"

In [10]:
def label_pol_third(row):
    if str(row['category1']).lower() == 'politics' or str(row['category2']).lower() == 'politics' or str(row['category3']).lower() == 'politics':
        return 1
    elif not any('politics' in str(category).lower() for category in row[['category1', 'category2', 'category3', 'category4', 'category5', 'category6', 'category7', 'category8', 'category9', 'category10']]):
        return 0
    else:
        return 2

In [ ]:
pd.options.mode.chained_assignment = None
df['is_politics_top_3'] = df.apply(label_pol_third, axis=1)

In [12]:
df.columns

Index(['potentialOutPath', 'transcript', 'wCount', 'rssUrl', 'epTitle',
       'epDescription', 'duration', 'pubDate', 'copyright', 'itunes:type',
       'itunes:complete', 'guid', 'itunes:explicit', 'enclosure',
       'itunes:image', 'transDict', 'id', 'title', 'lastUpdate', 'link',
       'lastHttpStatus', 'dead', 'contentType', 'itunesId', 'originalUrl',
       'itunesAuthor', 'itunesOwnerName', 'explicit', 'imageUrl', 'itunesType',
       'generator', 'newestItemPubdate', 'language', 'oldestItemPubdate',
       'episodeCount', 'popularityScore', 'priority', 'createdOn',
       'updateFrequency', 'chash', 'host', 'newestEnclosureUrl', 'podcastGuid',
       'podDescription', 'category1', 'category2', 'category3', 'category4',
       'category5', 'category6', 'category7', 'category8', 'category9',
       'category10', 'newestEnclosureDuration', 'oldestItemDatetime',
       'cleanDates', 'cleanDatesLoc', 'parsedDuration', 'hours', 'transEnts',
       'transStarts', 'transEnds', 'trans

In [14]:
df[df["is_politics_top_3"]==1].shape

(7792, 74)

## By using this strategy, we have 7792 transcripts as "politics". That's enough data to start with

In [15]:
df_pol = df[df["is_politics_top_3"]==1].copy()

In [16]:
df_pol.to_csv("pol_trans.csv", index=False)